pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller_test.go

```diff
@@ -696,6 +701,148 @@ func TestAttemptToScaleDown(t *testing.T) {
+				cm := wellKnownEtcdEndpointsConfigMap()
+				cm.Data["m-4"] = "10.0.139.81"
+				return []runtime.Object{cm}
+			}(),
+			fakeEtcdClientOptions: etcdcli.WithFakeClusterHealth(&etcdcli.FakeMemberHealth{Healthy: 3, Unhealthy: 1}),
+			expectedError:         fmt.Errorf("cannot proceed with scaling down, unhealthy voting members found: [https://10.0.139.78:1234]"),
+			validateFn: func(t *testing.T, fakeEtcdClient etcdcli.EtcdClient) {
+				memberList, err := fakeEtcdClient.MemberList(context.TODO())
+				if err != nil {
+					t.Fatal(err)
+				}
+				if len(memberList) != 4 {
+					t.Errorf("expected exactly 4 members, got %v", len(memberList))
+				}
+			},
+		},
+		{
+			name:                       "scale down only by one machine at a time when more than one machine pending deletion while quorum maintained",
+			initialObservedConfigInput: wellKnownReplicasCountSet,
+			initialEtcdMemberList: func() []*etcdserverpb.Member {
+				members := append(wellKnownEtcdMemberList(), &etcdserverpb.Member{
+					Name:     "m-4",
+					ID:       4,
+					PeerURLs: []string{"https://10.0.139.81:1234"},
+				}, &etcdserverpb.Member{
+					Name:     "m-5",
+					ID:       5,
+					PeerURLs: []string{"https://10.0.139.82:1234"},
+				})
+				return members
+			}(),
+			initialObjectsForMachineLister: func() []runtime.Object {
+				m4 := machineWithHooksFor("m-4", "10.0.139.81")
+				m5 := machineWithHooksFor("m-5", "10.0.139.82")
+				machines := wellKnownMasterMachines()
+				// override the default health config for test scenario
+				m1, ok := machines[0].(*machinev1beta1.Machine)
+				if !ok {
+					t.Fatalf("expected type *machinev1beta1.Machine, but got %T instead", m1)
+				}
+				m1.DeletionTimestamp = &metav1.Time{}
+				m2, ok := machines[1].(*machinev1beta1.Machine)
+				if !ok {
+					t.Fatalf("expected type *machinev1beta1.Machine, but got %T instead", m2)
+				}
+				m2.DeletionTimestamp = &metav1.Time{}
+				machines = append(machines, m4, m5)
+				return machines
+			}(),
+			initialObjectsForConfigMapTargetNSLister: func() []runtime.Object {
+				cm := wellKnownEtcdEndpointsConfigMap()
+				cm.Data["m-4"] = "10.0.139.81"
+				cm.Data["m-5"] = "10.0.139.82"
+				return []runtime.Object{cm}
+			}(),
+			fakeEtcdClientOptions: etcdcli.WithFakeClusterHealth(&etcdcli.FakeMemberHealth{Healthy: 3, Unhealthy: 2}),
+			validateFn: func(t *testing.T, fakeEtcdClient etcdcli.EtcdClient) {
+				memberList, err := fakeEtcdClient.MemberList(context.TODO())
+				if err != nil {
+					t.Fatal(err)
+				}
+				if len(memberList) != 4 {
+					t.Errorf("expected exactly 4 members, got %v", len(memberList))
+				}
+
+				// expected m1 to be deleted, m2 not to be deleted
+				var m1Found, m2Found bool
+				for _, member := range memberList {
+					if member.ID == 1 {
+						m1Found = true
+					}
+					if member.ID == 2 {
+						m2Found = true
+					}
+				}
+				if m1Found {
+					t.Fatal("expected m1 to be deleted")
```

Comment:
why m1? m-4 and m-5 are unhealthy and deleted

Reply:
the healthFake sets the first `n` members to unhealthy, where `n` is a number of `unhealthy` argument. 

instead of changing the healthFake to have control which members is healthy/unhealthy, I just override it for my test case

Reply:
ah right, and then we just test that m1 get's deleted first and m2 stays for the time being. 

Thanks for the explanation!